Sign and send tx via metamask

KarishmaBothara opened the pull request:
Details - This PR features metamask integration via portal.

Update extrinsic page, staking page as well react-signer

https://user-images.githubusercontent.com/29415595/214206648-a6908e3b-12c1-49aa-8998-90b345222d7b.mov

ken-centrality commented on packages/react-hooks/src/useMetaMask.ts:
```diff
@@ -0,0 +1,80 @@
+  const wallet = useWeb3React();
+  const provider = wallet?.provider;
+  const [isMetaMask, setIsMetaMask] = useState<boolean | undefined>();
+  const chainId = 3999;
```
there is no need for `chainId` here as we only use MetaMask as signer. But if we do, then we need to figure out how to adjust that depending on what Network we are switching to

ken-centrality commented on packages/react-hooks/src/useMetaMask.ts:
```diff
@@ -0,0 +1,80 @@
+
+import { initializeConnector, useWeb3React } from "@web3-react/core";
+import { MetaMask } from "@web3-react/metamask";
+import { useAtom } from "jotai";
```
could be overkill, is there an existing state management here?

aidan-starke replied:
I don't think it's needed here at all, it was just being used on `app-hub` so user can "disconnect" their account

KarishmaBothara replied:
just used this one from root-app-hub.. can remove it

ken-centrality commented on packages/react-signer/src/TxMetamaskSign.tsx:
```diff
@@ -0,0 +1,148 @@
+// Copyright 2017-2023 @polkadot/react-signer authors & contributors
```
rename to `TxMetaMaskSign.tsx`

ken-centrality reviewed (commented):
- Sign button is stuck when I cancel the signing in MetaMask

aidan-starke commented on packages/page-extrinsics/src/Submission.tsx:
```diff
@@ -60,17 +83,37 @@ function Selection ({ className, defaultValue }: Props): React.ReactElement<Prop
+        value={signMethod}
+        withLabel={true}
       />
+      { !signMethod ? <></> :
```
Can probably remove this line, or I think it could be better to populate `InputAddress` with `wallet.account` here and disable the dropdown if `signMethod === 'MetaMask'

aidan-starke commented on packages/page-staking/src/Actions/partials/Bond.tsx:
```diff
@@ -105,6 +123,24 @@ function Bond ({ className = '', isNominating, minNominated, minNominatorBond, m
+          value={signMethod}
+          withLabel={true}
+        />
+        { !signMethod ? <></> :
```
```suggestion

```

aidan-starke reviewed (commented):
Might be outside of the scope of this PR, but I think we need to find a way to populate `Accounts` page from MetaMask

KarishmaBothara commented:
> Might be outside of the scope of this PR, but I think we need to find a way to populate `Accounts` page from MetaMask

ya...good to have it as another PR after we merge this and pull upstream changes

KarishmaBothara commented:
> * Sign button is stuck when I cancel the signing in MetaMask

resolved

KarishmaBothara commented:
> @KarishmaBothara is this PR ready to be reviewed after the first round of feedback? I cannot seem to use the removal of `jotai` package even though the conversation is marked as resolved.
> 
> One more thing, not sure if I had mentioned, we don't need to trigger network switching in MetaMask, as we only use it for signing, not sending

sorry..my bad.. I ended up addressing all the review comments on spike/duplicate_MetaMaskIntegration branch
<img width="1360" alt="Screenshot 2023-01-31 at 7 38 54 AM" src="https://user-images.githubusercontent.com/29415595/215657149-627a680a-1b9e-4256-b2b4-76cd9caa8c68.png">

will update it on spike/MetaMaskIntegration branch

KarishmaBothara commented:
PR ready to be reviewed

ken-centrality reviewed (commented):
New changes look good.

One small change I would like to have is to set MetaMask as the default Signer, is it doable?

![Screenshot 2023-02-01 at 16 14 26](https://user-images.githubusercontent.com/97480229/215936695-59ccaa6e-cc5b-4169-bd8c-61c1af33de78.png)

KarishmaBothara commented:
> New changes look good.
> 
> One small change I would like to have is to set MetaMask as the default Signer, is it doable?
> 
> ![Screenshot 2023-02-01 at 16 14 26](https://user-images.githubusercontent.com/97480229/215936695-59ccaa6e-cc5b-4169-bd8c-61c1af33de78.png)

yes... should be doable.. will do it today

KarishmaBothara commented:
> set MetaMask as the default Signer

done!!
